content/post/elm-at-rakuten/index.md

```diff
@@ -701,42 +882,59 @@ Some consider this to be a benefit as it brings you out of your comfort zone and
 
 But for someone else is a burden and could discourage a team from adopting Elm.
 
-## 5. Content-driven static websites
+[Index](#index) | [Top](#top)
 
-Elm is not a good fit to build static websites that are mostly content-driven. In these cases, an old server-side rendered website can be a better option.
-
-On the other side, if you get to like Elm, it is hard to go back to plain JavaScript/HTML/CSS, so we experimented with Elm for static websites. There are several tools for static site generation. We used *elm-starter*[^elm-starter], a library that transforms an Elm website into a PWA that is also installable, works off-line and works without JavaScript.
-
-[^elm-starter]: These are the most common tools to generate static sites in Elm: [elm-pages](https://package.elm-lang.org/packages/dillonkearns/elm-pages/latest/) | [elmstatic](https://github.com/alexkorban/elmstatic) | [elm-starter](https://github.com/lucamug/elm-starter).
-
-The library *elm-starter* pre-renders pages at build time, making it also a candidate for SEO. The results so far are good, and the Lighthouse scores are promising.
+---
 
-## 6. Some JavaScript and CSS are still necessary
+## 5. Some JavaScript and CSS are still necessary
 
-Ideally, we could build an application by just writing in the Elm language. But if you need to use a third-party library not converted to Elm, we still need to use JavaScript. Doing so means entering again in the realm of possible runtime errors.
+Ideally, we could build an application by just writing in the Elm language. But if you need to use a third-party library not converted to Elm, we still need to use JavaScript. Doing so means entering again into the realm of possible runtime errors.
 
 Elm provides three ways to interact with external libraries: Flags, Ports, and Custom Elements.[^interop] All of them require you to write some JavaScript.
 
 In our case, for example, we must use a JavaScript library for handling payments.
 
-The required CSS, while using **Elm-UI**, is limited. In our applications, we have small snippets of CSS that are mainly tricks to support IE11.
+The required CSS, while using the library _Elm-UI_, is limited. In our applications, we have small snippets of CSS that are mainly tricks to support IE11.
 
 Related to this, Elm is probably not a good fit for short projects that require lots of integration with third-party JavaScript libraries.
 
 [^interop]: Elm provides several methodologies to communicate with JavaScript. Here is [an introduction to JavaScript interoperability](https://guide.elm-lang.org/interop/) with some [examples](https://github.com/elm-community/js-integration-examples).
 
```

Comment:
I would put a `6. Limited Polymorphic Abstraction Power`. At least this is a personal issue for me.
To put it shortly, in JavaScript you would use (or abuse) duck typing (i.e. dynamic polymorphism) to do many different things. A lot of these things are better served by the simple type-safe facilities Elm offers like sum types and first order generics. But some of these things cannot be done with the type-safe abstraction Elm offers.

I think the easiest example (and something I ran into when working on Flipper) is a multi-page app. My memory might be a bit off, but each page had its own TEA and therefore its own model, view and update. The main TEA is responsible for routing and converting all the messages, models and commands between the pages. The problem is that this ends up in adding a lot of boilerplate for any page you add in multiple different places. It is quite easy to forget to add something and your new page will not work (it is quite hard to make a mistake in all the boring boilerplate code — the Elm type-system will generally protect you against that).
I could shave off some of this boilerplate by using _explicit type-classes_ (basically the approach detailed in [Scrap Your Type Classes](https://www.haskellforall.com/2012/05/scrap-your-type-classes.html)). This improves the situation a little bit, but still requires boilerplate.

We could have saved some of the boilerplate if Elm had a concept of type classes, or even just plain interfaces (as in Java). Another source of boilerplate (dealing with different model types and message types and converting between them) is something that could have been reduced by Higher-kinded Types or Associated Types.

The rebuttal for all of the above (and why I would still use Elm in many projects) is that the JS _does allow_ for more abstraction power, but at the same time it doesn't provide any standardized method or system of doing these abstraction. With JS we almost always end up with a big hard-to-understand tangle of abstractions created by different developers who had their own idea about the right abstraction. TypeScript gives a more type-safe way to define these abstractions, but doesn't alleviate the issue.

I keep thinking Elm is following the celebrated Go philosophy much better than Go. The much-hyped Go language design philosophy is ostensibly "Keep the language simple and you'll prevent developers from using the wrong abstraction and creating hard-to-maintain code". The problem is that Go has many complex features (magic built-in types like channels, slices, arrays and maps, multiple returns, the way channels work) that should have been part of the library. Go also forces you to make all your data mutable and makes closures extremely cumbersome to use (on purpose), so you end up with something that resembles the Algol of the 1960s or the Pascal of the 1970s. I thought this philosophy is a sham, until I started using Elm and for the first time in my life I was fine with trading some abstraction power for simplicity. I'm still not 100% sold on this approach, but I can understand the benefits.

I'm not sure how to phrase all of that nicely, especially since my Elm is a little rusty. But just my thoughts on what could perhaps be a future article.

Reply:
@lucamug Please treat this as a comment since it will be too big to write... You can just resolve this thing :)

Reply:
@boazy , thank you for the feedback.

The main issue mentioned about Flipper I think it comes from inexperience (it was one of our first Elm project) and lack of tooling. 

About the boilerplate, at the moment is quite standard to use the library `elm-spa` (https://www.elm-spa.dev/) to build Single Page Applications. The library that take care of most of the wiring and allows us to decided if a page need the TEA or is just a view.

The topic of typeclasses has been debated a lot in the Elm community (for example https://discourse.elm-lang.org/search?q=typeclasses) and it is still unclear to me how much they would improve the language. I think it could be a topic for another post.